README.md
Rentpath Autotagging Plugin
___________________________
# jquery-autotagging
Client-side interface for Rentpath warehouse

## Setup

1. `bundle install`
2. `bower install`
3. `npm install`
4. `gulp <task>`, default watches `src/*.coffee` and builds `/dist`


## Tests

`rake jasmine`

Tests are currently using the compiled modules in the `/dist/shared` directory

## Upgrading

`WH#bindBodyClicked()` was removed. I don't think any client code used that method, but I'm mentioning it here, just in case.

When you upgrade to version 2 of this library, remove it from the `paths` section of your RequireJS configuration file and add it to the `packages` section..


## Examples

### Tag/Pixel Firing - [demo - click me](http://wh.consumersource.com/wtd.gif?site=www.qa.apartmentguide.com&site_version=www.qa.apartmentguide.com_kilo&cg=home&path=%2F&ft=4040.750000043772&type=pageview&cb=0&sess=1401052257136.1408034329918&fpc=1401052257136&title=Apartments%20for%20Rent%20-%20Your%20Trusted%20Apartment%20Finder%20Tool%20at%20ApartmentGuide.com&bs=960x679&sr=1101x1713&os=Mac&browser=Chrome&ver=36&ref=&registration=0&person_id=JWDykWnpFFPs4HDP7JPY36w9Xip&ad_sense_channel=1747283222&zutron=%5Bobject%20Object%5D&refinements=%5Bobject%20Object%5D&search_criteria=%5Bobject%20Object%5D&site_optimization=%5Bobject%20Object%5D&listingMediaCache=%5Bobject%20Object%5D&user_id=JWDykWnpFFPs4HDP7JPY36w9Xip)

### Requirejs - [example](examples/index.html)

### NonAMD - [example](examples/non_amd.html)
A standard js file built from our AMD modules that can be used in applications that do not use requirejs.
Everything is packaged and scoped except **jQuery**.

The module is accessible through the window variable `jquery_autotagging`.


## API

Some of these keys (like site_version) can be passed in on initialization to override a params default value.

| param | name | usage - description |
| ------------- |:-------------:| -----:|
| cg | Content group | Taken from a meta tag in the page, e.g SearchResults, HomePage, etc |
| sg  | Subgroup | Describes a group/widget/area on the page, e.g refinements, frontpageproperty |
| item  | Group | Used for grouping links such as distance links |
| value | Distinct value | Usually the text the user sees such as "10 miles from center". This can be overridden by setting a data attribute of "autotag-value". |
| type  | type | Indicates whether the event was a pageview beacon or a click event This can either be "pageview" or "click" |
| cb  | cache buster | cache busting value that will increment once per event |
| sess | session | A unique user identifier for a given session This is joined on a dot with the visit identifier (timestamp) |
| fpc | First party cookie | Just the user identifier from above This is a separate value so as to eliminate back-end string splitting/parsing |
| site  | Domain |  e.g www.apartmentguide.com |
| path  | path | The path of the current page, plus any possible query string parameters |
| title | title | HTML Title of the current page |
| bs  | Browser Size | dimensions of the user's window, e.g 1024x768 |
| sr  | Screen Resolution|  dimensions of the user's entire screen |
| os | Operating System |  The operating system of the user (experimental) |
| browser | browser | The web browser name (experimental) |
| ver | Browser Version | The browser version string (experimental) |
| ref | Referral |The referrer |
| firstVisit | First Visit|A timestamp that will be fired as part of the page-view event when a new session is initiated |
| tu  | truncated | unknown |
| tlc | Total Listing Count | count returned from search |
| spg | All Listing ids| Lisiting ids shown on search result page separated by semicolons |
| lfpos | Total Results | @results.records_per_page * (@results.page_number - 1) |
| lpp | Total Results |@results.records_per_page |
| dpg | Listing id | Listing id shown on detail page |
| hw  | Hardware | (mobile only) the hardware make and model |
| flashValue  | Adobe Flash Version |(mobile only) (Android only) the version of flash on the device |
| city  | city | the city from the search term |
| state | state | the state from the search term |
| zip | zip | the zip from the search term |
| registration| registration | unknown |
| ft  |fired time| unknown |
| site_version | site_version | the version of the site displayed to the user, used when the site dynamically scaled for the device (eg #{domain}_(nano|deca|kilo) |

## Data Attributes

Any data attributes that may affect autotagging should be prefixed with "autotag". E.g., adding "data-autotag-value='foo'" on an element will override the default "value" setting.

## Warning
`site_version` for ag_sites should always have domain set to microsites.com as requested by BI


CHANGELOG.md
# v2.1.1
* [deps] Use `browser-detect` rather than `browserdetect`.

[Compare v2.1.0..v2.1.1](https://github.com/RentPath/jquery-autotagging/compare/v2.1.0...v2.1.1)


# v2.1.0
* [breaking change] Remove calls to event.stopPropagation() and
  event.preventDefault(). [#26]
* [feature] Callbacks no longer waiting until the tag is loaded. Reduces
  page load times by at least 100ms. [#26]
* [feature] Fire multiple tags at once. Internally it no longer re-uses
  the same <img> element. [#26]

[Compare v2.0.4..v2.1.0](https://github.com/RentPath/jquery-autotagging/compare/v2.0.4...v2.1.0)


spec/javascripts/AutotaggingSpec.js
describe("Autotagging Suite", function() {
  var wh, simulPlatform, testDocument, testWindow;

  beforeEach(function() {
    var ready = false;

    require(['../../dist/shared/jquery.autotagging'], function(WH) {
      wh = new WH();
      wh.platform = {OS: 'OS', browser: 'dummy', version: ''};
      testWindow = $('<div></div>');
      ready = true;
    });

    waitsFor(function(){
      return ready;
    });
  });

  describe("Instance Methods", function() {
    describe('#determineWindowDimensions', function() {
      it('returns a string for window dimensions', function() {
        testWindow.width(100).height(100);
        expect(wh.determineWindowDimensions(testWindow)).toEqual('100x100');
      });
    });

    describe("#fire", function() {
      var callback;
      var obj;

      beforeEach(function() {
        obj = {};
        wh.init()
        callback = {
          fightFire: function(obj) {
            obj.blaze = true;
            return obj;
          }
        };

        wh.fireCallback = callback.fightFire;
        spyOn(wh, 'fireCallback').andCallThrough();
      });

      it('calls the given callback', function () {
        wh.fire(obj);
        expect(obj.blaze).toEqual(true);
        expect(wh.fireCallback).toHaveBeenCalled();
      });

      it('amends the object with the one time data ONCE', function() {
        wh.setOneTimeData({auxiliary: 'secret'});
        wh.fire(obj);
        expect(obj.auxiliary).toEqual('secret');

        next = {};
        wh.fire(next);
        expect(next.auxiliary).toEqual(undefined);
      });

      describe('records campaign id', function() {
        afterEach(function() {
          $.removeCookie('campaign_id');
        });

        it("returns the user's campaign ID if a campaign cookie is set", function() {
          $.cookie('campaign_id', '1234', {path: '/'});
          wh.fire(obj);
          expect(obj.campaign_id).toEqual('1234');
        });

        it("does not return the user's campaign ID if a campaign cookie is not present", function() {
          wh.fire(obj);
          expect(obj.campaign_id).toBeUndefined();
        });
      });

      describe('records login information', function() {
        var signon_cookie;
        var provider_cookie;
        var zid_cookie;

        beforeEach(function() {
          provider_cookie = $.cookie('provider');
          signon_cookie   = $.cookie('sgn');
          zid_cookie      = $.cookie('zid');
        });

        afterEach(function() {
          $.cookie('provider', provider_cookie, {expires: 3650, path: '/'});
          $.cookie('sgn', signon_cookie, {expires: 3650, path: '/'});
          $.cookie('zid', zid_cookie, {expires: 3650, path: '/'});
        });

        it('returns "0" if not registered', function() {
          $.cookie('sgn', 0, {expires: 3650, path: '/'});
          wh.fire(obj);
          expect(obj.registration).toEqual(0);
        });

        it('returns "1" if registered', function() {
          $.cookie('sgn', 1, {expires: 3650, path: '/'});
          wh.fire(obj);
          expect(obj.registration).toEqual(1);
        });
      });

      it('sets the firing time', function () {
        wh.fire(obj);
        expect(obj.ft).toNotEqual(undefined);
      });
    });

    describe('#firstClass', function() {
      it('yields the first class name of the element', function() {
        testElement = $("<div class='first second third'></div>");
        expect(wh.firstClass(testElement)).toEqual('first');
      });
    });

    describe('#getDataFromMetaTags', function() {
      it('extracts WH meta tags', function() {
        testDoc = $("<div><meta name='WH.cg' content=''/><meta name='WH.test' content='dummy'/><meta name='WH.quiz' content='placeholder'</div>");
        result = { cg : '', test : 'dummy', quiz : 'placeholder' };
        expect(wh.getDataFromMetaTags(testDoc)).toEqual(result);
      });
    });

    describe("#getSubgroupId", function() {
      it('yields the id of the first parent element', function() {
        loadFixtures("autotagging.html")
        link = $("a:contains('Browse New Homes')");
        expect(wh.getSubgroupId(link)).toEqual('nav_menu');
      });

      it('yields null when no id present', function() {
        loadFixtures("autotagging.html")
        link = $("html");
        expect(wh.getSubgroupId(link)).toEqual(null);
      });
    });

    describe("#getItemId", function() {
      it('yields the id of the element', function() {
        testElement = $("<div id='foo'></div>");
        expect(wh.getItemId(testElement)).toEqual('foo');
      });

      it('yields the first class of the element when no id present', function() {
        testElement = $("<div class='first second third'></div>");
        expect(wh.getItemId(testElement)).toEqual('first');
      });
    });

    describe("#init", function() {
      var targets;

      beforeEach(function() {
        setFixtures('<div id="nav_menu"><span class="icon_home sprite">Stuff</span><a class="trap" href="#">Privacy Policy</a><a class="x">X</a></div>');
        targets = 'a.trap, span.icon_home';
        wh.init({clickBindSelector: targets});
        spyOn(wh, 'fire');
      });

      it('binds to the named elements', function() {
        $(document).find('a.trap').click();
        expect(wh.fire).toHaveBeenCalled();
      });

      it('binds to multiple elements', function() {
        $(document).find('span.icon_home').click();
        expect(wh.fire).toHaveBeenCalled();
      });

      it('does not bind to other elements', function() {
        $(document).find('a.x').click();
        expect(wh.fire).not.toHaveBeenCalled();
      });

      describe ("with config object", function() {
        var configObject, opts;

        beforeEach(function() {
          configObject = {cg: 'test', lpp: "1"};
          opts = {metaData: configObject};
        });

        it('should use the config object', function() {
          wh.init(opts);
          expect(wh.metaData).toEqual(configObject);
        });
      });

      describe ("without config object", function() {
        it('should fallback to meta tags when config object is missing', function() {
          spyOn(wh, 'getDataFromMetaTags')
          wh.init();
          expect(wh.getDataFromMetaTags).toHaveBeenCalled();
        });
      });
    });

    describe("default clickBindSelector", function() {
      var newContent;

      beforeEach(function() {
        setFixtures("<div><input type=submit><input type=button><a class='trap' href='#'>O</a><img src='http://www.example.com' title='Image'><a class='x' href='#'>O</a></div>");
        targets = 'a.trap, img, input[type=submit], input[type=button]';
        wh.init({clickBindSelector: targets});
        spyOn(wh, 'fire');
      });

      it('binds to the named elements', function() {
        $(document).find('a.trap').click();
        expect(wh.fire).toHaveBeenCalled();
      });

      it('binds to img elements', function() {
        $(document).find('img').click();
        expect(wh.fire).toHaveBeenCalled();
      });

      it('binds to input submit elements', function() {
        $(document).find('input[type=submit]').click();
        expect(wh.fire).toHaveBeenCalled();
      });

      it('binds to input button elements', function() {
        $(document).find('input[type=button]').click();
        expect(wh.fire).toHaveBeenCalled();
      });
    });

    describe("#getSessionID", function() {
      var time = 123;

      beforeEach(function () {
        $.cookie('WHSessionID','');
      });

      it('uses the cookie when present', function() {
        $.cookie('WHSessionID',111222);
        expect(wh.getSessionID(time)).toEqual('111222');
      });

      it('uses the time when cookie is null', function() {
        spyOn($, 'cookie').andReturn(null);
        expect(wh.getSessionID(time)).toEqual(time);
      });

      it('uses the time when cookie is undefined', function() {
        spyOn($, 'cookie').andReturn(undefined);
        expect(wh.getSessionID(time)).toEqual(time);
      });

      it('sets firstVisit to time when cookie is null', function() {
        spyOn($, 'cookie').andReturn(null);
        wh.getSessionID(time)
        expect(wh.firstVisit).toEqual(time);
      });
    });

    describe('#setOneTimeData', function() {
      it('records attributes', function() {
        once = {a: 'Apple', b: 'Banana'};
        wh.setOneTimeData(once);
        data = wh.getOneTimeData();
        expect(data.a).toEqual('Apple');
        expect(data.b).toEqual('Banana');
      });
    });

    describe("followHref", function() {
      it('defaults to true', function() {
        wh.init();
        expect(wh.followHref).toEqual(true);
      });

      it('overrides default with argument', function() {
        wh.init({followHref: false});
        expect(wh.followHref).toEqual(false);
      });
    });

    describe("#desktopOrMobile", function() {
      it('recognizes desktop', function() {
        expect(wh.desktopOrMobile(1025)).toEqual('kilo');
      });
      it('recognizes tablets', function() {
        expect(wh.desktopOrMobile(768)).toEqual('deca');
      });
     it('recognizes mobile', function() {
        expect(wh.desktopOrMobile(767)).toEqual('nano');
      });
    });

    describe("#determineReferrer", function() {
      beforeEach(function() {
        testDocument = $('<div></div>');
        testDocument.referrer = "rawr";
        testWindow = $('<div></div>');
        testWindow.location = {
          href: ""
        };
      });

      it('should use real_referrer when use_real_referrer is true', function() {
        $.cookie('real_referrer', 'woof');
        testWindow.location.href = "http://www.rentpathsite.com/?use_real_referrer=true"
        expect(wh.determineReferrer(testDocument, testWindow)).toEqual("woof");
      });

      it('should use document.referrer when use_real_referrer is false', function() {
        $.cookie('real_referrer', 'woof');
        testWindow.location.href = "http://www.rentpathsite.com/?use_real_referrer=false"
        expect(wh.determineReferrer(testDocument, testWindow)).toEqual("rawr");
      });
    });

    describe("#replaceDoubleByteChars", function() {
      it('should replace double-byte chars', function() {
        wh.init();
        expect(wh.replaceDoubleByteChars("Download Android™ App©")).toEqual("Download Android(tm) App(c)");
      });
    });

    describe('#eventHandlers', function() {
      it('should set an instance variable for backwards compatibility', function() {
        wh.eventHandlers({});
        expect(wh.clickHandler).toBeDefined();
      });
    });
  });
});


spec/javascripts/config/bootstrap.js
var root = '../../../';
require.config({
  shim: {},
  paths: {
    browser-detect: root + 'vendor/javascripts/browser-detect/dist/browser-detect',
    jquery: root + 'vendor/javascripts/jquery/dist/jquery',
    underscore: root + 'vendor/javascripts/underscore/underscore',
    'jquery.cookie': root + 'vendor/javascripts/jquery.cookie/jquery.cookie',
    requirejs: root + 'vendor/javascripts/requirejs/require'
  }
});


click_handler.js
// Generated by CoffeeScript 1.8.0
(function() {
  var __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  define(['jquery'], function($) {
    var ClickHandler;
    ClickHandler = (function() {
      function ClickHandler(wh, opts) {
        this.wh = wh;
        if (opts == null) {
          opts = {};
        }
        this.elemClicked = __bind(this.elemClicked, this);
        this.clickBindSelector = opts.clickBindSelector || 'a, input[type=submit], input[type=button], img';
        this.dataAttributePrefix = opts.dataAttributePrefix || 'autotag';
        if (opts.exclusions != null) {
          this.clickBindSelector = this.clickBindSelector.replace(/,\s+/g, ":not(" + opts.exclusions + "), ");
        }
      }

      ClickHandler.prototype.bind = function(elem) {
        return $(elem).on('click', this.clickBindSelector, this.elemClicked);
      };

      ClickHandler.prototype._shouldRedirect = function(href) {
        return (href != null) && (href.indexOf != null) && href.indexOf('javascript:') === -1;
      };

      ClickHandler.prototype._followHrefConfigured = function(event, options, wh) {
        var _ref, _ref1;
        if ((event != null ? (_ref = event.data) != null ? _ref.followHref : void 0 : void 0) != null) {
          return event != null ? (_ref1 = event.data) != null ? _ref1.followHref : void 0 : void 0;
        } else if ((options != null ? options.followHref : void 0) != null) {
          return options != null ? options.followHref : void 0;
        } else if ((wh != null ? wh.followHref : void 0) != null) {
          return wh != null ? wh.followHref : void 0;
        } else {
          return false;
        }
      };

      ClickHandler.prototype._setDocumentLocation = function(href) {
        return document.location = href;
      };

      ClickHandler.prototype._openNewWindow = function(href) {
        return window.open(href);
      };

      ClickHandler.prototype.elemClicked = function(e, options) {
        var attr, attrs, domTarget, getClosestAttr, href, item, jQTarget, realName, subGroup, target, trackingData, value, _i, _len, _ref;
        if (options == null) {
          options = {};
        }
        domTarget = e.target;
        attrs = domTarget.attributes;
        jQTarget = $(e.target);
        if (!jQTarget.is(this.clickBindSelector)) {
          jQTarget = jQTarget.parent();
        }
        item = this.wh.getItemId(jQTarget) || '';
        subGroup = this.wh.getSubgroupId(jQTarget) || '';
        value = this.wh.replaceDoubleByteChars(jQTarget.data("" + this.dataAttributePrefix + "-value") || jQTarget.text()) || '';
        trackingData = {
          sg: subGroup,
          item: item,
          value: value,
          type: 'click',
          x: e.clientX,
          y: e.clientY
        };
        for (_i = 0, _len = attrs.length; _i < _len; _i++) {
          attr = attrs[_i];
          if (attr.name.indexOf('data-') === 0 && (_ref = attr.name, __indexOf.call(this.wh.exclusionList, _ref) < 0)) {
            realName = attr.name.replace('data-', '');
            trackingData[realName] = attr.value;
          }
        }
        getClosestAttr = function(attr) {
          return jQTarget.attr(attr) || jQTarget.closest('a').attr(attr);
        };
        href = getClosestAttr('href');
        target = getClosestAttr('target');
        if (this._followHrefConfigured(e, options, this.wh) && this._shouldRedirect(href)) {
          e.preventDefault();
          if ((target === "_blank") || e.ctrlKey || e.metaKey) {
            this._openNewWindow(href);
          } else {
            trackingData.afterFireCallback = (function(_this) {
              return function() {
                return _this._setDocumentLocation(href);
              };
            })(this);
          }
        }
        this.wh.fire(trackingData);
        return e.stopPropagation();
      };

      return ClickHandler;

    })();
    return ClickHandler;
  });

}).call(this);


select_change_handler.js
// Generated by CoffeeScript 1.8.0
(function() {
  var __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  define(['jquery'], function($) {
    var SelectChangeHandler;
    SelectChangeHandler = (function() {
      function SelectChangeHandler(wh) {
        this.wh = wh;
        this.recordChange = __bind(this.recordChange, this);
      }

      SelectChangeHandler.prototype.bind = function(doc) {
        return $(doc).on('change', 'select', this.recordChange);
      };

      SelectChangeHandler.prototype.recordChange = function(evt) {
        var $target, attributes, domTarget;
        domTarget = evt.target;
        attributes = domTarget.attributes;
        $target = $(evt.target);
        this.wh.fire(this.trackingData(evt, $target, attributes, this.wh), $target);
        return evt.stopPropagation();
      };

      SelectChangeHandler.prototype.item = function($target, wh) {
        return this.value($target, wh);
      };

      SelectChangeHandler.prototype.subgroup = function($target, wh) {
        return wh.getSubgroupId($target) || '';
      };

      SelectChangeHandler.prototype.value = function($target, wh) {
        var value;
        value = $target.find(':selected').val();
        return wh.replaceDoubleByteChars(value) || '';
      };

      SelectChangeHandler.prototype.text = function($target, wh) {
        var text;
        text = $target.find(':selected').text();
        return wh.replaceDoubleByteChars(text);
      };

      SelectChangeHandler.prototype.trackingData = function(evt, $target, attributes, wh) {
        var attribute, data, realName, _i, _len, _ref;
        data = {
          sg: this.subgroup($target, wh),
          item: this.item($target, wh),
          value: this.value($target, wh),
          text: this.text($target, wh),
          type: 'change',
          x: evt.clientX,
          y: evt.clientY
        };
        for (_i = 0, _len = attributes.length; _i < _len; _i++) {
          attribute = attributes[_i];
          if (attribute.name.indexOf('data-') === 0 && (_ref = attribute.name, __indexOf.call(this.wh.exclusionList, _ref) < 0)) {
            realName = attribute.name.replace('data-', '');
            data[realName] = attribute.value;
          }
        }
        return data;
      };

      return SelectChangeHandler;

    })();
    return SelectChangeHandler;
  });

}).call(this);


jquery.autotagging.js
(function() {
  var __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  define(['jquery', 'browser-detect', 'underscore', './click_handler', './select_change_handler', 'jquery.cookie'], function($, browserdetect, _, ClickEventHandler, SelectChangeHandler) {
    var WH;
    return WH = (function() {
      var DESKTOP_WIDTH, MOBILE_WIDTH;

      function WH() {
        this.obj2query = __bind(this.obj2query, this);
        this.firedTime = __bind(this.firedTime, this);
        this.fire = __bind(this.fire, this);
        this.elemClicked = __bind(this.elemClicked, this);
        this.clearOneTimeData = __bind(this.clearOneTimeData, this);
        this.init = __bind(this.init, this);
      }

      WH.prototype.WH_SESSION_ID = 'WHSessionID';

      WH.prototype.WH_LAST_ACCESS_TIME = 'WHLastAccessTime';

      WH.prototype.WH_USER_ID = 'WHUserID';

      WH.prototype.THIRTY_MINUTES_IN_MS = 30 * 60 * 1000;

      WH.prototype.TEN_YEARS_IN_DAYS = 3650;

      MOBILE_WIDTH = 768;

      DESKTOP_WIDTH = 1023;

      WH.prototype.cacheBuster = 0;

      WH.prototype.domain = '';

      WH.prototype.firstVisit = null;

      WH.prototype.metaData = null;

      WH.prototype.oneTimeData = null;

      WH.prototype.path = '';

      WH.prototype.performance = window.performance || {};

      WH.prototype.sessionID = '';

      WH.prototype.userID = '';

      WH.prototype.warehouseTag = null;

      WH.prototype.charMap = {
        8482: '(tm)',
        169: '(c)',
        174: '(r)'
      };

      WH.prototype.init = function(opts) {
        var handler, _i, _len, _ref, _results;
        if (opts == null) {
          opts = {};
        }
        this.domain = document.location.host;
        this.setSiteVersion(opts);
        this.exclusionList = opts.exclusionList || [];
        this.fireCallback = opts.fireCallback;
        this.path = "" + document.location.pathname + document.location.search;
        this.warehouseURL = opts.warehouseURL;
        this.opts = opts;
        this.followHref = opts.followHref != null ? opts.followHref : true;
        this.setCookies();
        this.determineDocumentDimensions(document);
        this.determineWindowDimensions(window);
        this.determinePlatform(window);
        opts.metaData || (opts.metaData = {});
        _.extend(opts.metaData, this.getDataFromMetaTags(document));
        this.metaData = opts.metaData;
        this.firePageViewTag();
        _ref = this.eventHandlers(opts);
        _results = [];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          handler = _ref[_i];
          _results.push(handler.bind(document));
        }
        return _results;
      };

      WH.prototype.clearOneTimeData = function() {
        return this.oneTimeData = void 0;
      };

      WH.prototype.getSubgroupId = function(elem) {
        var closestId;
        closestId = elem.closest('[id]').attr('id');
        return closestId || null;
      };

      WH.prototype.determineWindowDimensions = function(obj) {
        obj = $(obj);
        return this.windowDimensions = "" + (obj.width()) + "x" + (obj.height());
      };

      WH.prototype.determineDocumentDimensions = function(obj) {
        obj = $(obj);
        return this.browserDimensions = "" + (obj.width()) + "x" + (obj.height());
      };

      WH.prototype.determinePlatform = function(win) {
        return this.platform = browserdetect.platform(win);
      };

      WH.prototype.determineReferrer = function(doc, win) {
        if (win.location.href.match(/\?use_real_referrer\=true/)) {
          return $.cookie('real_referrer');
        } else {
          return doc.referrer;
        }
      };

      WH.prototype.elemClicked = function(e, options) {
        if (options == null) {
          options = {};
        }
        return this.clickHandler.elemClicked(e, options);
      };

      WH.prototype.setSiteVersion = function(opts) {
        if (opts.metaData) {
          return this.siteVersion = "" + (opts.metaData.site_version || this.domain) + "_" + (this.deviceType());
        } else {
          return this.siteVersion = "" + this.domain + "_" + (this.deviceType());
        }
      };

      WH.prototype.deviceType = function() {
        return this.device || (this.device = this.desktopOrMobile());
      };

      WH.prototype.desktopOrMobile = function(deviceWidth) {
        if (deviceWidth == null) {
          deviceWidth = $(window).width();
        }
        switch (false) {
          case !this.desktop(deviceWidth):
            return 'kilo';
          case !this.tablet(deviceWidth):
            return 'deca';
          case !this.mobile(deviceWidth):
            return 'nano';
        }
      };

      WH.prototype.desktop = function(deviceWidth) {
        return deviceWidth > DESKTOP_WIDTH;
      };

      WH.prototype.tablet = function(deviceWidth) {
        return deviceWidth >= MOBILE_WIDTH && deviceWidth <= DESKTOP_WIDTH;
      };

      WH.prototype.mobile = function(deviceWidth) {
        return deviceWidth < MOBILE_WIDTH;
      };

      WH.prototype.fire = function(obj, $element) {
        var key;
        obj.ft = this.firedTime();
        obj.cb = this.cacheBuster++;
        obj.sess = "" + this.userID + "." + this.sessionID;
        obj.fpc = this.userID;
        obj.site = this.domain;
        obj.path = this.path;
        obj.title = $('title').text();
        obj.bs = this.windowDimensions;
        obj.sr = this.browserDimensions;
        obj.os = this.platform.OS;
        obj.browser = this.platform.browser;
        obj.ver = this.platform.version;
        obj.ref = obj.ref || this.determineReferrer(document, window);
        obj.registration = $.cookie('sgn') === '1' ? 1 : 0;
        if ($.cookie('sgn') != null) {
          obj.person_id = $.cookie('zid');
        }
        if ($.cookie('campaign_id') != null) {
          obj.campaign_id = $.cookie('campaign_id');
        }
        obj.site_version = this.siteVersion;
        if (obj.site_version != null) {
          this.metaData.site_version = obj.site_version;
        }
        if (obj.cg != null) {
          this.metaData.cg = obj.cg;
        }
        if (this.metaData.cg == null) {
          this.metaData.cg = '';
        }
        if (typeof this.fireCallback === "function") {
          this.fireCallback(obj);
        }
        if (this.oneTimeData != null) {
          for (key in this.oneTimeData) {
            obj[key] = this.oneTimeData[key];
          }
          this.clearOneTimeData();
        }
        if (this.firstVisit) {
          obj.firstVisit = this.firstVisit;
          this.firstVisit = null;
        }
        return this.obj2query($.extend(obj, this.metaData), (function(_this) {
          return function(query) {
            var requestURL;
            requestURL = _this.warehouseURL + query;
            if (requestURL.length > 2048 && navigator.userAgent.indexOf('MSIE') >= 0) {
              requestURL = requestURL.substring(0, 2043) + "&tu=1";
            }
            if (!_this.warehouseTag) {
              _this.warehouseTag = $('<img/>', {
                id: 'PRMWarehouseTag',
                border: '0',
                width: '1',
                height: '1'
              });
            }
            $element = $element || $('body');
            _this.warehouseTag.unbind('load').load(function() {
              return $element.trigger('WH_pixel_success_' + obj.type);
            });
            _this.warehouseTag.unbind('error').error(function() {
              return $element.trigger('WH_pixel_error_' + obj.type);
            });
            if (obj.afterFireCallback) {
              _this.warehouseTag.unbind('load').unbind('error').bind('load', obj.afterFireCallback).bind('error', obj.afterFireCallback);
            }
            return _this.warehouseTag[0].src = requestURL;
          };
        })(this));
      };

      WH.prototype.firedTime = function() {
        var now;
        now = this.performance.now || this.performance.webkitNow || this.performance.msNow || this.performance.oNow || this.performance.mozNow;
        return ((now != null) && now.call(this.performance)) || new Date().getTime();
      };

      WH.prototype.firePageViewTag = function(options) {
        if (options == null) {
          options = {};
        }
        options.type = 'pageview';
        return this.fire(options);
      };

      WH.prototype.getItemId = function(elem) {
        return elem.attr('id') || this.firstClass(elem);
      };

      WH.prototype.firstClass = function(elem) {
        var klasses;
        if (!(klasses = elem.attr('class'))) {
          return;
        }
        return klasses.split(' ')[0];
      };

      WH.prototype.getDataFromMetaTags = function(obj) {
        var metaTag, metas, name, retObj, _i, _len;
        retObj = {};
        metas = $(obj).find('meta');
        for (_i = 0, _len = metas.length; _i < _len; _i++) {
          metaTag = metas[_i];
          metaTag = $(metaTag);
          if (metaTag.attr('name') && metaTag.attr('name').indexOf('WH.') === 0) {
            name = metaTag.attr('name').replace('WH.', '');
            retObj[name] = metaTag.attr('content');
          }
        }
        return retObj;
      };

      WH.prototype.getOneTimeData = function() {
        return this.oneTimeData;
      };

      WH.prototype.sort_order_array = ["site", "site_version", "firstvisit", "tu", "cg", "listingid", "dpg", "type", "sg", "item", "value", "ssSiteName", "ssTestName", "ssVariationGroupName", "spg", "lpp", "path", "logged_in", "ft"];

      WH.prototype.setTagOrder = function(obj) {
        var elem, index, key, prop_key_array, result_array;
        prop_key_array = [];
        result_array = [];
        for (key in obj) {
          prop_key_array.push(key);
        }
        for (elem in this.sort_order_array) {
          index = $.inArray(this.sort_order_array[elem], prop_key_array);
          if (index > 0) {
            result_array.push(prop_key_array[index]);
            prop_key_array.splice(index, 1);
          }
        }
        result_array = result_array.concat(prop_key_array);
        return result_array;
      };

      WH.prototype.obj2query = function(obj, cb) {
        var elem, key, rv, tag_order, val;
        tag_order = this.setTagOrder(obj);
        rv = [];
        for (elem in tag_order) {
          key = tag_order[elem];
          if (obj.hasOwnProperty(key) && ((val = obj[key]) != null)) {
            rv.push("&" + key + "=" + (encodeURIComponent(val)));
          }
        }
        cb(rv.join('').replace(/^&/, '?'));
      };

      WH.prototype.getSessionID = function(currentTime) {
        if ($.cookie(this.WH_SESSION_ID) != null) {
          return $.cookie(this.WH_SESSION_ID);
        } else {
          this.firstVisit = currentTime;
          return this.firstVisit;
        }
      };

      WH.prototype.setCookies = function() {
        var sessionID, timestamp, userID;
        userID = $.cookie(this.WH_USER_ID);
        timestamp = (new Date()).getTime();
        if (!userID) {
          userID = timestamp;
          $.cookie(this.WH_USER_ID, userID, {
            expires: this.TEN_YEARS_IN_DAYS,
            path: '/'
          });
        }
        sessionID = this.getSessionID(timestamp);
        $.cookie(this.WH_SESSION_ID, sessionID, {
          path: '/'
        });
        $.cookie(this.WH_LAST_ACCESS_TIME, timestamp, {
          path: '/'
        });
        this.sessionID = sessionID;
        return this.userID = userID;
      };

      WH.prototype.setOneTimeData = function(obj) {
        var key, _results;
        this.oneTimeData || (this.oneTimeData = {});
        _results = [];
        for (key in obj) {
          _results.push(this.oneTimeData[key] = obj[key]);
        }
        return _results;
      };

      WH.prototype.replaceDoubleByteChars = function(str) {
        var char, result;
        result = (function() {
          var _i, _len, _ref, _results;
          _ref = str.split('');
          _results = [];
          for (_i = 0, _len = _ref.length; _i < _len; _i++) {
            char = _ref[_i];
            _results.push(this.charMap[char.charCodeAt(0)] || char);
          }
          return _results;
        }).call(this);
        return result.join('');
      };

      WH.prototype.eventHandlers = function(options) {
        var selectChangeHandler;
        this.clickHandler = new ClickEventHandler(this, options);
        selectChangeHandler = new SelectChangeHandler(this);
        return [this.clickHandler, selectChangeHandler];
      };

      return WH;

    })();
  });

}).call(this);
